addons/web/static/src/services/router_service.js

```diff
@@ -55,105 +55,125 @@ function getRoute() {
   return { pathname, search: searchQuery, hash: hashQuery };
 }
 
-/**
- * @param {function} getCurrent function that returns the current route
- * @returns {function} function to compute the next hash
- */
-export function makePreProcessQuery(getCurrent) {
-  const lockedKeys = new Set();
-  return (hash) => {
-    const newHash = {};
-    Object.keys(hash).forEach((key) => {
-      if (lockedKeys.has(key)) {
-        return;
-      }
-      const k = key.split(" ");
-      let value;
-      if (k.length === 2) {
-        value = hash[key];
-        key = k[1];
-        if (k[0] === "lock") {
-          lockedKeys.add(key);
-        } else if (k[0] === "unlock") {
-          lockedKeys.delete(key);
-        } else {
-          return;
-        }
-      }
-      newHash[key] = value || hash[key];
-    });
-    const current = getCurrent();
-    Object.keys(current.hash).forEach((key) => {
-      if (lockedKeys.has(key) && !(key in newHash)) {
-        newHash[key] = current.hash[key];
-      }
-    });
-    return newHash;
-  };
+function applyLocking(lockedKeys, hash, currentRoute, lock=false) {
+  const newHash = {};
+  Object.keys(hash).forEach((key) => {
+    if (!lock && lockedKeys.has(key)){ // forbid implicit override of key
+      return;
+    }
+    if (lock === 'on') {
+      lockedKeys.add(key);
+    } else if (lock === 'off') {
+      lockedKeys.delete(key);
+    }
+    newHash[key] = hash[key];
+  });
+
+  Object.keys(currentRoute.hash).forEach((key) => {
+    if (lockedKeys.has(key) && !(key in newHash)) {
+      newHash[key] = currentRoute.hash[key];
+    }
+  });
+  return newHash;
 }
 
-function makeRouter(env) {
+export function routerSkeleton(env, {getRoute, historyPush, historyReplace, redirect}) {
   let bus = env.bus;
   let current = getRoute();
-  window.addEventListener("hashchange", () => {
+  const lockedKeys = new Set();
+
+  bus.on('hashchange', null, () => {
     current = getRoute();
-    bus.trigger("ROUTE_CHANGE");
+    bus.trigger('ROUTE_CHANGE');
```

Comment:
so we basically trigger 2 events on env.bus each time hashchange is triggered on window? Would be nice to remove the noise (the first event could be private I think)